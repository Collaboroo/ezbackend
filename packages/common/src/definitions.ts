import {EzBackend} from '@ezbackend/core'
import {Connection, ConnectionOptions} from 'typeorm'
import {FastifyInstance, FastifyLoggerOptions} from "fastify"
import { EzModelMeta } from '.'

export interface IOptions {
    server?: {
        port?: number
        logger?:FastifyLoggerOptions
    }
    orm?: ConnectionOptions
    
}

declare module '@ezbackend/core' {
    interface EzBackend {
        orm: Connection,
        server: FastifyInstance
        models: Array<EzModelMeta>
    }
}

