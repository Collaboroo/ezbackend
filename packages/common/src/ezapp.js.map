{"version":3,"file":"ezapp.js","sourceRoot":"","sources":["ezapp.ts"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,wCAAkD;AAElD,kEAA+B;AAS/B,SAAS,0BAA0B,CAC/B,MAAa,EACb,QAAyC;IAGzC,OAAO;QAAC,cAAe;aAAf,UAAe,EAAf,qBAAe,EAAf,IAAe;YAAf,yBAAe;;QACnB,MAAM,CAAC,SAAS,CAAC,IAAI;QACjB,+DAA+D;QAC/D,UAAC,MAAuB,IAAK,OAAC,MAAM,CAAC,QAAQ,CAAC,OAAhB,MAAM,EAAsB,IAAI,GAAjC,CAAkC,CAClE,CAAA;IACL,CAAC,CAAA;AACL,CAAC;AAED,2CAA2C;AAC3C,wCAAwC;AACxC,SAAS,YAAY,CAAC,MAAa;IAC/B,OAAO;QACH,kDAAkD;QAClD,QAAQ;QACR,MAAM,EAAE,0BAA0B,CAAoD,MAAM,EAAE,QAAQ,CAAC;QACvG,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,IAAI,EAAE,0BAA0B,CAAkD,MAAM,EAAE,MAAM,CAAC;QACjG,KAAK,EAAE,0BAA0B,CAAmD,MAAM,EAAE,OAAO,CAAC;QACpG,IAAI,EAAE,0BAA0B,CAAkD,MAAM,EAAE,MAAM,CAAC;QACjG,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,OAAO,EAAE,0BAA0B,CAAqD,MAAM,EAAE,SAAS,CAAC;QAC1G,GAAG,EAAE,0BAA0B,CAAiD,MAAM,EAAE,KAAK,CAAC;QAC9F,KAAK,EAAE,0BAA0B,CAAmD,MAAM,EAAE,OAAO,CAAC;QACpG,mJAAmJ;QACnJ,OAAO,EAAE,0BAA0B,CAAmD,MAAM,EAAE,SAAS,CAAC;QACxG,SAAS,EAAE,0BAA0B,CAAmD,MAAM,EAAE,WAAW,CAAC;QAC5G,qBAAqB,EAAE,0BAA0B,CAAmE,MAAM,EAAE,uBAAuB,CAAC;QACpJ,oBAAoB,EAAE,0BAA0B,CAAkE,MAAM,EAAE,sBAAsB,CAAC;QACjJ,QAAQ,EAAE,0BAA0B,CAAsD,MAAM,EAAE,UAAU,CAAC;QAC7G,aAAa,EAAE,0BAA0B,CAA2D,MAAM,EAAE,eAAe,CAAC;QAC5H,eAAe,EAAE,0BAA0B,CAA6D,MAAM,EAAE,iBAAiB,CAAC;QAClI,MAAM,EAAE,0BAA0B,CAAgD,MAAM,EAAE,QAAQ,CAAC;QACnG,6EAA6E;QAC7E,QAAQ,EAAE,0BAA0B,CAAM,MAAM,EAAE,UAAU,CAAoB;QAChF,kBAAkB,EAAE,0BAA0B,CAAgE,MAAM,EAAE,oBAAoB,CAAC;QAC3I,eAAe,EAAE,0BAA0B,CAA6D,MAAM,EAAE,iBAAiB,CAAC;KAErI,CAAA;AACL,CAAC;AAID;;;GAGG;AACH;IAA2B,yBAAG;IAM1B;;OAEG;IACH;QAAA,YACI,iBAAO,SASV;QAjBS,gBAAU,GAAoB,EAAE,CAAA;QAmB1C,6BAA6B;QAC7B,kFAAkF;QAClF,YAAM,GAAG,0BAA0B,CAAoD,KAAI,EAAE,QAAQ,CAAC,CAAA;QACtG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,UAAI,GAAG,0BAA0B,CAAkD,KAAI,EAAE,MAAM,CAAC,CAAA;QAChG,WAAK,GAAG,0BAA0B,CAAmD,KAAI,EAAE,OAAO,CAAC,CAAA;QACnG,UAAI,GAAG,0BAA0B,CAAkD,KAAI,EAAE,MAAM,CAAC,CAAA;QAChG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,aAAO,GAAG,0BAA0B,CAAqD,KAAI,EAAE,SAAS,CAAC,CAAA;QACzG,SAAG,GAAG,0BAA0B,CAAiD,KAAI,EAAE,KAAK,CAAC,CAAA;QAC7F,WAAK,GAAG,0BAA0B,CAAmD,KAAI,EAAE,OAAO,CAAC,CAAA;QACnG,aAAO,GAAG,0BAA0B,CAAM,KAAI,EAAE,SAAS,CAA+B,CAAA;QACxF,eAAS,GAAG,0BAA0B,CAAmD,KAAI,EAAE,WAAW,CAAC,CAAA;QAC3G,2BAAqB,GAAG,0BAA0B,CAAmE,KAAI,EAAE,uBAAuB,CAAC,CAAA;QACnJ,0BAAoB,GAAG,0BAA0B,CAAkE,KAAI,EAAE,sBAAsB,CAAC,CAAA;QAChJ,cAAQ,GAAG,0BAA0B,CAAsD,KAAI,EAAE,UAAU,CAAC,CAAA;QAC5G,mBAAa,GAAG,0BAA0B,CAA2D,KAAI,EAAE,eAAe,CAAC,CAAA;QAC3H,qBAAe,GAAG,0BAA0B,CAA6D,KAAI,EAAE,iBAAiB,CAAC,CAAA;QAEjI,cAAQ,GAAG,0BAA0B,CAAsD,KAAI,EAAE,UAAU,CAAC,CAAA,CAAC,yCAAyC;QACtJ,wBAAkB,GAAG,0BAA0B,CAAgE,KAAI,EAAE,oBAAoB,CAAC,CAAA;QAC1I,qBAAe,GAAG,0BAA0B,CAA6D,KAAI,EAAE,iBAAiB,CAAC,CAAA;QACjI,mEAAmE;QACnE,sDAAsD;QAEtD,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACvH,aAAO,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,OAAO,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACjH,iBAAW,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,WAAW,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACzH,mBAAa,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,aAAa,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC7H,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACvH,oBAAc,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,cAAc,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC/H,eAAS,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,SAAS,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QACrH,YAAM,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,MAAM,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QAC/G,gBAAU,GAAG,UAAC,QAAgB,EAAE,MAA+C,IAAM,iBAAM,UAAU,aAAC,QAAQ,EAAC,MAAM,CAAC,CAAA,CAAA,CAAC,CAAA;QA3CnH,KAAI,CAAC,UAAU,CAAC,oBAAoB,EAAE,UAAO,QAAQ,EAAE,IAAI;;gBACvD,QAAQ,CAAC,MAAM,GAAG,YAAY,CAAC,IAAI,CAAC,CAAA;;;aACvC,CAAC,CAAA;QACF,KAAI,CAAC,cAAc,CAAC,oBAAoB,EAAE,UAAO,QAAQ,EAAE,IAAI;;gBAC3D,uGAAuG;gBACvG,YAAY;gBACZ,OAAO,QAAQ,CAAC,MAAM,CAAA;;;aACzB,CAAC,CAAA;;IACN,CAAC;IAfD,sBAAI,4BAAS;aAAb,cAAkB,OAAO,IAAI,CAAC,UAAU,CAAA,CAAC,CAAC;;;OAAA;IAoD1C;;;;OAIG;IACH,sCAAsB,GAAtB,UAAuB,MAAuB,EAAE,MAAa;QAA7D,iBAiBC;QAfG,IAAM,SAAS,GAAG,UAAO,MAAuB;;gBAC5C,MAAM,CAAC,SAAS,CAAC,OAAO,CAAC,UAAA,IAAI;oBACzB,IAAI,CAAC,MAAM,CAAC,CAAA;gBAChB,CAAC,CAAC,CAAA;gBACF,MAAM,CAAC,IAAI,CAAC,OAAO,CAAC,UAAA,GAAG;oBACnB,IAAI,GAAG,YAAY,KAAK,EAAE;wBACtB,GAAG,CAAC,sBAAsB,CAAC,MAAM,EAAE,GAAG,CAAC,CAAA;qBAC1C;gBACL,CAAC,CAAC,CAAA;;;aACL,CAAA;QAED,+GAA+G;QAC/G,IAAM,eAAe,GAAG,CAAC,MAAM,CAAC,KAAK,KAAK,kBAAW,CAAC,MAAM,CAAC,CAAC,CAAC,CAAC,IAAA,wBAAE,EAAC,SAAS,CAAC,CAAC,CAAC,CAAC,SAAS,CAAA;QAEzF,MAAM,CAAC,QAAQ,CAAC,eAAe,EAAE,MAAM,CAAC,IAAI,CAAC,CAAA;IACjD,CAAC;IACL,YAAC;AAAD,CAAC,AA/ED,CAA2B,UAAG,GA+E7B;AA/EY,sBAAK"}