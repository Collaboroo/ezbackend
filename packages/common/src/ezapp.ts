import { App } from "@ezbackend/core"

function generateFastifyFuncWrapper(parent,funcName:string) {

    return (...opts) => {
        parent._functions.push(
            (server) => server[funcName](...opts)
        )
    }
}

function createServer(parent:EzApp) {
    return {
        route: generateFastifyFuncWrapper(parent,'route'),
        register: generateFastifyFuncWrapper(parent,'register'),
        addSchema: generateFastifyFuncWrapper(parent,'addSchema'),
        decorateReply: generateFastifyFuncWrapper(parent,'decorateReply'),
        inject: generateFastifyFuncWrapper(parent,'inject'),
        addHook: generateFastifyFuncWrapper(parent,'addHook'),
    }
}

export class EzApp extends App {

    protected _functions: Array<Function> = []

    get functions() { return this._functions }

    constructor() {
        super()
        this.setHandler("Create Server Stub", async (instance, opts) => {
            instance.server = createServer(this)
        })
        this.setPostHandler("Remove Server Stub", async (instance, opts) => {
            delete instance.server
        })

    }

    registerFastifyPlugins(server,parent) {

        server.register(async (server, opts) => {
            parent.functions.forEach(func => {
                func(server)
            })
            parent.apps.forEach(app => {
                if (app instanceof EzApp) {
                    app.registerFastifyPlugins(server,app)
                }
            })
        }, parent.opts)
    }
}