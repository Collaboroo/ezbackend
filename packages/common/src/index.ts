import { IOptions } from "./definitions";
import { IEzbConfig } from "@ezbackend/core";
import { fastify } from "fastify";
import path from "path";
import { EzBackend } from "./definitions";
import { mixedInstance } from "avvio";
import { createConnection } from "typeorm";
import { RouteOptions } from "fastify";
import { APIGenerator } from "./models";
import {convert} from './models/typeorm-json-schema'

const ezb = EzBackend.app();

//Configure defaults
ezb.plugins.init = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig & IOptions,
  cb
) => {
  ezb.orm = await createConnection(opts.orm);
  ezb.server = fastify(opts.server);
  cb();
};

ezb.plugins.handler = (ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  const customEzbPath =
    opts.entryPoint ?? path.join(process.cwd(), ".ezb/index.ts");
  require(customEzbPath);
  ezb.models.forEach((model) => {
    const repo = ezb.orm.getRepository(model);

    //LEFT OFF
    // const metaData = ezb.orm.getMetadata(model)
    // console.log(convert(metaData))

    const generator = new APIGenerator(repo, { prefix: repo.metadata.name });
    generator.generateRoutes();
  });
  cb();
};

ezb.plugins.run = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig,
  cb
) => {
  await ezb.server.listen(opts.port, function (err, address) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
  cb();
};

export { EzBackend } from "./definitions";
export { EzModel, response, APIGenerator } from "./models";
