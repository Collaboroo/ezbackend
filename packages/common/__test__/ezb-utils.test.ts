import { afterAll, beforeAll, describe, expect, test } from "@jest/globals";

import { EzBackend } from '../src'
import ezb from "./test.index"

describe("Test Utils", () => {

    it("Should throw an error if the instance has not started and you try to use printRoutes, printPlugins or prettyPrint", async () => {

        const app = new EzBackend()

        const errored = {
            printRoutes: false,
            printPlugins: false,
            prettyPrint: false
        }
        try {
            app.printRoutes()
        } catch (e) {
            errored.printRoutes = true
        }
        try {
            app.printPlugins()
        } catch (e) {
            errored.printPlugins = true
        }
        try {
            app.prettyPrint()
        } catch (e) {
            errored.prettyPrint = true
        }
        expect(errored).toEqual({
            printRoutes: true,
            printPlugins: true,
            prettyPrint: true
        })
    })

    describe("Functionality test", () => {
        beforeAll(async () => {
            await ezb.start({
                port: 3000,
                backend: {
                    fastify: {
                        logger: false
                    },
                    typeorm: {
                        database: ':memory:'
                    }
                }
            })


        });

        afterAll(async () => {
            const instance = ezb.getInternalInstance()
            await instance.orm.close();
            await instance._server.close();
        });

        it("Should be able to print routes", async () => {

            //Injection to initialize fastify to view the routes
            await ezb.inject({
                method: "GET",
                url: "/"
            })

            expect(ezb.printRoutes()).toMatchSnapshot()
        })

        it("Should be able to print all fastify plugins nicely", async () => {

            //Injection to initialize fastify to view the routes
            await ezb.inject({
                method: "GET",
                url: "/"
            })

            const fastifyPluginsWithoutLoadTimings = ezb.printPlugins().replace(/\d+ ms/gm, "")

            expect(fastifyPluginsWithoutLoadTimings).toMatchSnapshot()
        })

        it("Should be able to print the plugin tree for debugging", async () => {
            //Injection to initialize fastify to view the routes
            await ezb.inject({
                method: "GET",
                url: "/"
            })
            //TODO: Make it able to easily obtain the plugin names
            const pluginsWithoutLoadTimings = ezb.prettyPrint().replace(/\d+ ms/gm, "")

            expect(pluginsWithoutLoadTimings).toMatchSnapshot()

        })

    })


})