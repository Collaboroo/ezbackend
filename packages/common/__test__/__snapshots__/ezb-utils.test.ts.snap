// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Utils Functionality test Should be able to print all fastify plugins nicely 1`] = `
"bound root 
├── bound _after 
├─┬ childFunc 
│ ├── ezbErrorPage 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ ├─┬ childFunc 
│ │ ├─┬ childFunc 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ ├── bound _after 
│ │ │ └── bound _after 
│ │ └── childFunc 
│ └─┬ childFunc 
│   ├─┬ childFunc 
│   │ ├── bound _after 
│   │ ├── bound _after 
│   │ ├── bound _after 
│   │ ├── bound _after 
│   │ └── bound _after 
│   └── childFunc 
├─┬ plugin 
│ ├── bound _after 
│ └── bound _after 
└── bound _after 
"
`;

exports[`Test Utils Functionality test Should be able to print routes 1`] = `
"└── /
    ├── Sample (POST)
    │   Sample (GET)
    │   ├── Nullable (POST)
    │   │   Nullable (GET)
    │   │   └── / (POST)
    │   │       / (GET)
    │   │       └── :id (GET)
    │   │           :id (PATCH)
    │   │           :id (DELETE)
    │   ├── / (POST)
    │   │   / (GET)
    │   │   └── :id (GET)
    │   │       :id (PATCH)
    │   │       :id (DELETE)
    │   └── Unique (POST)
    │       Unique (GET)
    │       └── / (POST)
    │           / (GET)
    │           └── :id (GET)
    │               :id (PATCH)
    │               :id (DELETE)
    ├── User (POST)
    │   User (GET)
    │   ├── 2 (POST)
    │   │   2 (GET)
    │   │   └── / (POST)
    │   │       / (GET)
    │   │       └── :id (GET)
    │   │           :id (PATCH)
    │   │           :id (DELETE)
    │   └── / (POST)
    │       / (GET)
    │       └── :id (GET)
    │           :id (PATCH)
    │           :id (DELETE)
    ├── NoCascade
    │   ├── User (POST)
    │   │   User (GET)
    │   │   └── / (POST)
    │   │       / (GET)
    │   │       └── :id (GET)
    │   │           :id (PATCH)
    │   │           :id (DELETE)
    │   └── Program (POST)
    │       Program (GET)
    │       └── / (POST)
    │           / (GET)
    │           └── :id (GET)
    │               :id (PATCH)
    │               :id (DELETE)
    ├── Detail (POST)
    │   Detail (GET)
    │   └── / (POST)
    │       / (GET)
    │       └── :id (GET)
    │           :id (PATCH)
    │           :id (DELETE)
    └── Program (POST)
        Program (GET)
        ├── 2 (POST)
        │   2 (GET)
        │   └── / (POST)
        │       / (GET)
        │       └── :id (GET)
        │           :id (PATCH)
        │           :id (DELETE)
        └── / (POST)
            / (GET)
            └── :id (GET)
                :id (PATCH)
                :id (DELETE)
"
`;

exports[`Test Utils Functionality test Should be able to print the plugin tree for debugging 1`] = `
"bound root 
├─┬ _init 
│ ├── Create Entities Container 
│ ├── Manage Event Subscriptions 
│ ├─┬ UnnamedApp1 
│ │ ├── Create \\"SampleNullable\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp2 
│ │ ├── Create \\"user\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp3 
│ │ ├── Create \\"noCascadeUser\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp4 
│ │ ├── Create \\"detail\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp5 
│ │ ├── Create \\"noCascadeProgram\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp6 
│ │ ├── Create \\"program\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp7 
│ │ ├── Create \\"sample\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp8 
│ │ ├── Create \\"SampleUnique\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ ├─┬ UnnamedApp9 
│ │ ├── Create \\"User2\\" Entity 
│ │ └─┬ rules 
│ │   └── Add Rules Subscriber 
│ └─┬ UnnamedApp10 
│   ├── Create \\"Program2\\" Entity 
│   └─┬ rules 
│     └── Add Rules Subscriber 
├─┬ _postInit 
│ ├── Create Database Connection 
│ ├─┬ UnnamedApp1 
│ │ └── Obtain SampleNullable Repository 
│ ├─┬ UnnamedApp2 
│ │ └── Obtain user Repository 
│ ├─┬ UnnamedApp3 
│ │ └── Obtain noCascadeUser Repository 
│ ├─┬ UnnamedApp4 
│ │ └── Obtain detail Repository 
│ ├─┬ UnnamedApp5 
│ │ └── Obtain noCascadeProgram Repository 
│ ├─┬ UnnamedApp6 
│ │ └── Obtain program Repository 
│ ├─┬ UnnamedApp7 
│ │ └── Obtain sample Repository 
│ ├─┬ UnnamedApp8 
│ │ └── Obtain SampleUnique Repository 
│ ├─┬ UnnamedApp9 
│ │ └── Obtain User2 Repository 
│ └─┬ UnnamedApp10 
│   └── Obtain Program2 Repository 
├─┬ _handler 
│ ├── Create Server Stub 
│ ├── Add Fastify Boom 
│ ├── Add SocketIO 
│ ├── Add Error Schema 
│ ├─┬ UnnamedApp1 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp2 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp3 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp4 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp5 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp6 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp7 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp8 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ ├─┬ UnnamedApp9 
│ │ ├── Create Server Stub 
│ │ ├─┬ router 
│ │ │ ├── Create Server Stub 
│ │ │ ├── Add Create Schema 
│ │ │ ├── Add Update Schema 
│ │ │ ├── Add Full Schema 
│ │ │ ├── Generate createOne route 
│ │ │ ├── Generate getOne route 
│ │ │ ├── Generate getAll route 
│ │ │ ├── Generate updateOne route 
│ │ │ └── Generate deleteOne route 
│ │ └─┬ rules 
│ │   └── Create Server Stub 
│ └─┬ UnnamedApp10 
│   ├── Create Server Stub 
│   ├─┬ router 
│   │ ├── Create Server Stub 
│   │ ├── Add Create Schema 
│   │ ├── Add Update Schema 
│   │ ├── Add Full Schema 
│   │ ├── Generate createOne route 
│   │ ├── Generate getOne route 
│   │ ├── Generate getAll route 
│   │ ├── Generate updateOne route 
│   │ └── Generate deleteOne route 
│   └─┬ rules 
│     └── Create Server Stub 
└─┬ _postHandler 
  ├── Remove Server Stub 
  ├── Create Fastify Server 
  ├── Attach Socket IO 
  ├── Register Fastify Plugins 
  ├── Add Request Context Plugin 
  ├── Set Request Context For Global Access 
  ├── Set SocketIO Context for global access 
  ├── Add middleware to authenticate outgoing packets 
  ├─┬ UnnamedApp1 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp2 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp3 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp4 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp5 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp6 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp7 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp8 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  ├─┬ UnnamedApp9 
  │ ├── Remove Server Stub 
  │ ├─┬ router 
  │ │ └── Remove Server Stub 
  │ └─┬ rules 
  │   └── Remove Server Stub 
  └─┬ UnnamedApp10 
    ├── Remove Server Stub 
    ├─┬ router 
    │ └── Remove Server Stub 
    └─┬ rules 
      └── Remove Server Stub 
"
`;
